packages/vite/src/node/plugins/importAnalysisBuild.ts

```diff
@@ -74,11 +74,23 @@ function preload(
       // @ts-ignore
       seen[dep] = true
       const isCss = dep.endsWith('.css')
-      const cssSelector = isCss ? '[rel="stylesheet"]' : ''
-      // @ts-ignore check if the file is already preloaded by SSR markup
-      if (document.querySelector(`link[href="${dep}"]${cssSelector}`)) {
-        return
-      }
+      const allLinks = Array.from(document.querySelectorAll('link'))
+      // When relativePreloadUrls is false then dep looks like /assets/foo.js
+      // and importerUrl === undefined so to get absolute URL with origin
+      // (https://blah.com/assets/foo.js) we need to provide location.href as a
+      // base to URL constructor. When relativePreloadUrls is true then we have
+      // importerUrl and dep is already converted to a full absolute URL by the
+      // assetsURL function
+      const depWithOrigin = new URL(dep, location.href).href
+      const isPreloaded = allLinks.some((link) => {
+        return (
+          // link.href is normalized by the browser to an absolute URL with
+          // origin even if link.getAttribute('href') is relative
+          link.href === depWithOrigin && (!isCss || link.rel === 'stylesheet')
+        )
+      })
+      if (isPreloaded) return
```

Comment:
When `relativePreloadUrls` is `false` (base isn't relative), then we don't have a bug, no? So we can leave the code as it was before. Or is there a bug here too? IIUC, the bug you reported only happens for relative base.
And in this case, what we need is to compare `new URL(link.href, location.href).href === dep`
I think the test cases may not be correctly testing the issue you reported because CI is passing, or maybe I'm missing something here.